docs/development: add section about codespell (#700)
kinvolk/lokomotive: docs/development/README.md

which is then passed to `codespell` via the `--ignore-words` flag.

If you find that some word you used should be ignored, add it to this file.

### Ignoring vendored files

Lokomotive ships some 3rd party code like vendored Go dependencies and Helm charts, which are maintained
by other people and may also contain some typos. It is recommended to configure `codespell` to **SKIP** those
files from checking, instead of **FIXING** them in our repository, to avoid deriving from upstream code, which
may make the update process of this code more difficult in the future (merge conflicts, extensive diffs etc.).

To skip some file or directory (recursively) from spell checking, add the path to `.codespell.skip` file.
The content of this file is then passed to the `--skip` flag of `codespell`.

Additionally, if time allows, consider fixing the found typos in the respective upstream projects.

### Spell checking git commit messages

Spell checking of git commit messages is not done as part of the CI process, but it is recommended to do it locally
before submitting patches.

To check for typos in commit messages of your feature branch, you can run the following command:
```sh

> git log master..HEAD | codespell -

Reviewer: This doesn't seem to work for me:

```
$ git show --quiet
commit 2f0e71cfc38ae5d9a63f8cd1be4f026ca53de961
Author: Iago López Galeiras <iago@kinvolk.io>
Date:   Tue Jul 7 17:33:17 2020 +0200

    Tis is a testz commit msgz

    I dont now how to wrait.
$ git log HEAD^1..HEAD | codespell -
$
```

Reviewer 2: @iaguis because it is valid as per codespell, try `echo "Tis is a testz commit msgz" | codespell -` it won't generate any error. And try `echo "Aks" | codespell -` it will

Reviewer: Hm, so `codespell` is not a full spell checker, it only checks for common misspellings.

Fair enough.

## Contents

- [Update component configuration.](./updating-components.md)

## Spell checking

> As part of CI process, we run [codespell](https://github.com/codespell-project/codespell) to find potential

Reviewer: ```suggestion
As part of the CI process, we run [codespell](https://github.com/codespell-project/codespell) to find potential
```

> typos in our code base. This section describes how to configure and use and it.

Reviewer 2: ```suggestion
typos in our code base. This section describes how to configure and use it.
```

Author: Applied, thanks :)

### Running locally

To run `codespell` locally, install it either using the [official method](https://github.com/codespell-project/codespell#installation)

> or use package manager specific to your Linux distribution.

Reviewer: ```suggestion
or using a package manager specific to your Linux distribution.
```

> or use package manager specific to your Linux distribution.

Reviewer 2: I think that the distribution can be any. I for one use it on macos

With `codespell` installed, run the following command to run spell checking locally:
```sh
make codespell
```

> It should find potential typos in the code.

Reviewer: I'd remove this sentence.

Author: Done.

> 
> It should find potential typos in the code.

Author: ```suggestion
```

### Ignoring specific words

> Sometimes, `codespell` will detect some words as typos, like `AKS`, `ACI` or `IAM`, which will cause the CI to

Reviewer: ```suggestion
Sometimes, `codespell` detects some words as typos, like `AKS`, `ACI` or `IAM`, which cause the CI to
```

> report errors. To resolve that, we define our list of ignored words in `.codespell.ignorewords` file,

Reviewer: ```suggestion
report errors. To resolve that, we define our list of ignored words in the `.codespell.ignorewords` file,
```

> which is then passed to `codespell` via `--ignore-words` flag.

Reviewer: ```suggestion
which is then passed to `codespell` via the `--ignore-words` flag.
```

> If you find that some word you added should be ignored, add it to this file.

Reviewer: ```suggestion
If you find that some word you used should be ignored, add it to this file.
```

### Ignoring vendored files

Lokomotive ships some 3rd party code like vendored Go dependencies and Helm charts, which are maintained
by other people and may also contain some typos. It is recommended to configure `codespell` to **SKIP** those
files from checking, instead of **FIXING** them in our repository, to avoid deriving from upstream code, which
may make the update process of this code more difficult in the future (merge conflicts, extensive diffs etc.).

> To skip some file or directory (recursively) from spell checking, add the path to `.codespell.skip` file.

Reviewer: ```suggestion
To recursively skip some file or directory from spell checking, add the path to `.codespell.skip` file.
```

Author: It implies "recursively skipping files", which does not make much sense to me :smile:

Reviewer: Fair enough, I'll resolve.

> The content of this file is then passed to `--skip` flag of `codespell`.

Reviewer: ```suggestion
The content of this file is then passed to the `--skip` flag of `codespell`.
```

> Additionally, if time allows, consider fixing found typos in the upstream projects.

Reviewer: ```suggestion
Additionally, if time allows, consider fixing the found typos in the respective upstream projects.
```

### Spell checking git commit messages

> Spell checking of git commit messages is not done as part of CI process, but it is recommended to do it locally
> before submitting the patches.

Reviewer: ```suggestion
Spell checking of git commit messages is not done as part of the CI process, but it is recommended to do it locally
before submitting patches.
```

To check for typos in commit messages of your feature branch, you can run the following command:
```sh

> git diff master..HEAD | grep -v ^- | codespell -

Reviewer: How does this work? AFAIK `git diff` doesn't show commit messages.

Author: Oops, I copied the wrong command from my script :smile: It suppose to be `git log master..HEAD | codespell -` :)

> git diff master..HEAD | grep -v ^- | codespell -

Author: ```suggestion
git log master..HEAD | codespell -
```

```

> It will spell check both commit message and added code, so it is recommended to first fix typos in the code before

Reviewer: ```suggestion
It will spell check both commit messages and new code, so it is recommended to first fix typos in the code before
```
